stub/src/main/java/io/grpc/stub/StreamObserver.java

```diff
@@ -43,9 +47,17 @@
    * server streaming calls, but may receive many onNext callbacks.  Servers may invoke onNext at
    * most once for client streaming calls, but may receive many onNext callbacks.
    *
-   * <p>If an exception is thrown by an implementation the caller is expected to terminate the
-   * stream by calling {@link #onError(Throwable)} with the caught exception prior to
-   * propagating it.
+   * <p>Although implementations are expected to not throw exceptions, if an exception occurs the
+   * caller is encouraged to call {@link #onError} as part of normal application exception handling.
+   * That is to say, {@code onError} should be called if any related part of the application throws
+   * an exception, generally independent of the source, to avoid leaks. gRPC will call {@code
+   * onError()} when an application's implementation throws.
+   *
+   * <p>As an exception to the rule and for historical reasons, on server-side, {@code onNext()} may
+   * throw with a {@code StatusRuntimeException} if the call is cancelled. This was necessary
+   * because there was not a callback available to notify the service. Services are encouraged to
```

Comment:
Consider omitting "This was necessary because ... " since teaching callers the actual historical reason doesn't seem worth the extra reading